### DISCUSSION TITLE
Best practices recommendation for Kotlin?

### DISCUSSION THREAD

**Discussion by barrettc** [JanusGraph/janusgraph#4706] [Q&A]
What, if any, are best practices recommendations when using the Java API from a Kotlin coroutine environment? Are there plans to support an asynchronous Java client in the future?

**Comment by porunov** [comment]
Gremlin Server supports async Gremlin queries scheduling ([see JavaDoc](https://tinkerpop.apache.org/javadocs/current/core/org/apache/tinkerpop/gremlin/driver/Client.html#submitAsync(java.lang.String))). Thus, you can use it already if needed.

If we talk about asynchronous driver with async API then it should be implemented on TinkerPop Gremlin side to allow Graph Providers to make their async implementations. As such, this is less async client is related to JanusGraph, but more related to Gremlin itself.

However, in modern Java I would look more into upgrading JanusGraph core to Java 24 / 25 and utilize Virtual Threads instead of adding async implementation. Usually, the main reason to use async capabilities was to reduce the amount of threads needed to sustain the load because Platform threads take up big chunk of RAM and we waste CPU cycles switching thread contexts. Previously people would schedule any blocking parts via async execution and return `Future<SomeResult>` and chain logic via `whenComplete -> doNext`. 
With Virtual Threads, however, we can easily block the code for any IO operations making async programming obsolete for such use-cases. Such a blocked thread won't take any CPU cycles because it won't be pinned to a Platform Thread and the RAM usage is negligible. In other words, we spin up as much threads as needed and they all work efficiently together. 
I didn't use Kotlin, so can't comment much about it, but I believe you can use CoroutineDispatcher backed by virtual threads. 

Perhaps, there could be other reasons where async programming could be more beneficial, so it might still make sense to add async GLV steps, but for this it would be nice to collect feedback why people need async API.
